Support for manual timing (see issue #198, https://github.com/google/benchmark/issues/198)

jknuuttila opened the pull request:
I have added support for manual timing, for benchmarking cases where CPU time and wallclock time are not useful. It is enabled by calling `UseManualTime` similar to how `UseRealTime` is used. When it is enabled, the benchmark loop must call `SetIterationTime(seconds)` once per each iteration. Manually timed benchmarks are decorated with "manual_time" when outputting results, and byte and item throughput are calculated with the manual time instead of the wallclock time.

I have added a test for manual timing in the `benchmark_test.cc` file. The test simulates a workload with `std::this_thread::sleep`, and measures the duration of the sleep manually using `std::chrono::high_resolution_clock::now`. The manually measured time is reported back with `SetIterationTime`. The benchmark is run with both `UseRealTime` and `UseManualTime` to see that the timings roughly match, which is expected in this case.

I have also added a short piece of documentation for the feature in the main `README.md` file.

I have also added one `static_cast` to replace an implicit cast which triggered a warning with Visual Studio 2015.

dominichamon commented on src/benchmark.cc:
```diff
@@ -647,7 +688,7 @@ void RunBenchmark(const benchmark::internal::Benchmark::Instance& b,
         }
         for (std::size_t ti = 0; ti < pool.size(); ++ti) {
-            pool[ti] = std::thread(&RunInThread, &b, iters, ti, &total);
+            pool[ti] = std::thread(&RunInThread, &b, iters, static_cast<int>(ti), &total);
```
nice catch. a shame we can't support more than 2 billion threads though :P

dominichamon commented on src/benchmark.cc:
```diff
@@ -699,7 +743,11 @@ void RunBenchmark(const benchmark::internal::Benchmark::Instance& b,
         // Report the total iterations across all threads.
         report.iterations = static_cast<int64_t>(iters) * b.threads;
-        report.real_accumulated_time = real_accumulated_time;
+        if (b.use_manual_time) {
```
do you think it's worth separating out manual accumulated time from real accumulated time? both could be set on a benchmark, and we already have both CPU and real accumulated time.

jknuuttila replied:
I'm not sure. Usually, for my use cases (GPU workloads), the real time doesn't make sense when manual time is required, and I would be interested in having the throughput measured against the manual time.

However, it's possible that there are cases where it could be useful to have both. The console output might get a little cluttered in tabular mode, but I guess that's not a big problem.

dominichamon replied:
I like to minimise surprises, and i think that if someone calls `->UseRealTime()->UseManualTime()` on a benchmark it should do the right thing (show real time and manual time).

yes, it might be cluttered in console mode but it's already fairly heavy-weight and the user has requested the extra data.

jknuuttila replied:
I can make a new version of the pull request that supports this.

jknuuttila commented:
I have implemented support for separate tracking of manual and real times.

However, this patch turned out to be very invasive and hairy, because it affects output formatting and stat computation significantly, and as a result it is very possible that bugs may have been introduced. Indeed, it seems that some automatic check failed (which I noticed when writing this comment), so there is certainly some bug.

After implementing this and seeing how many changes were required, I would now personally strongly prefer the original simpler approach, which is to just use the manual time instead of real-time (requiring no modifications to output and stat computation) when `UseManualTime` has been specified. Enabling both real time and manual time simultaneously seems like a very niche use case to begin with, and I don't believe it is worth the additional code complexity.

I would suggest handling the combination of `UseManualTime` and `UseRealTime` either with an error message (i.e. the combination is forbidden), or by executing the benchmark once in both modes and reporting them completely separately. If you agree, I can implement this functionality and submit it as a separate pull request. If you would still prefer tracking both separately, I can fix the bugs remaining in this pull request.

jknuuttila commented:
It seems that I also made some mistake while rebasing the pull request, and as a result commits of other people incorrectly show up as parts of this pull request. Perhaps it would be easiest if I just made a new pull request after it is decided which approach is the best one.

dominichamon commented:
alright.. start with the simpler approach (maybe you can just revert your local branch to that version, and avoid creating a new PR?) and then i may take a crack at extending it.

jknuuttila commented:
Sure! I'll try to revert the branch.

jknuuttila commented:
I updated the PR again with the simpler version, and this time I hopefully did the rebasing better, too.

I added a couple of CHECKs to warn against using real time and manual time simultaneously.

dominichamon commented on include/benchmark/benchmark_api.h:
```diff
@@ -444,6 +455,14 @@ class Benchmark {
+  // SetIterationTime(seconds) to report the measured time, which will be used
+  // to control how many iterations are run, and in the printing of items/second
+  // or MB/second values. If not called, either the CPU time or the elapsed time
+  // will be used (if UseRealTime() was called).
```
this would be an error now.

dominichamon replied:
update this comment, and i think we're good to go.

dominichamon commented on test/benchmark_test.cc:
```diff
@@ -174,5 +176,30 @@ static void BM_ParallelMemset(benchmark::State& state) {
+    state.SetIterationTime(elapsed.count());
+    slept_for += microseconds;
+  }
+  state.SetItemsProcessed(slept_for);
```
SetLabel might be a better method to use to report this. or just skip it .. it doesn't add anything to the example that i can see.

jknuuttila replied:
You mean the `SetItemsProcessed`? I put it there to test that the item throughput is calculated according to the manual time, as expected. I can remove it if you think it's a problem.

dominichamon replied:
ah, i see. it's fine :)

jknuuttila commented:
The comment should now be fixed.
